at_client/lib/src/service/sync_service_impl.dart

```diff
@@ -46,75 +50,112 @@ class SyncServiceImpl implements SyncService, AtSignChangeListener {
   }
 
   @override
-  Future<SyncResult> sync({Function? onDone, Function? onError}) async {
-    SyncResult syncResult;
-    // If sync in-progress, return.
-    if (_isSyncInProgress) {
-      _logger.info('Sync already in-progress. Cannot start a new sync');
-      syncResult = SyncResult();
-      syncResult.syncStatus = SyncStatus.failure;
-      syncResult.atClientException = AtClientException(
-          'AT0014', 'Sync-InProgress. Cannot start a new sync process');
-      if (onError != null) {
-        onError(syncResult);
-      }
-      return syncResult;
+  void sync({Function? onDone, Function? onError}) {
+    final syncRequest = SyncRequest();
+    syncRequest.onDone = onDone;
+    syncRequest.onError = onError;
+    syncRequest.requestedOn = DateTime.now().toUtc();
+    syncRequest.result = SyncResult();
+    _syncRequests.add(syncRequest);
+    //#TODO should we call _processQueue here or in cron or in _statsServiceListener ?
+    if (_syncRequests.length > _syncRequestThreshold ||
+        (_syncRequests.isNotEmpty &&
+            _syncRequests[0]
+                    .requestedOn
+                    .difference(DateTime.now().toUtc())
+                    .inSeconds >
+                _syncRequestTriggerInSeconds)) {
+      _logger.finer('triggering sync');
+      _processQueue();
     }
-    try {
-      //Setting isSyncInProgress to true to prevent parallel sync calls.
-      _isSyncInProgress = true;
-      // If network is not available, return.
+    return;
+  }
+
+  /// Listens on stats notification sent by the cloud secondary server
+  Future<void> _statsServiceListener() async {
+    _statsNotificationListener = await NotificationServiceImpl.create(_atClient)
+        as NotificationServiceImpl;
+    // Setting the regex to 'statsNotification' to receive only the notifications
+    // from stats notification service.
+    _statsNotificationListener
+        .subscribe(regex: 'statsNotification')
+        .listen((notification) async {
+      final serverCommitId = notification.value;
+      if (serverCommitId != null &&
+          int.parse(serverCommitId) > await _getLocalCommitId()) {
+        final syncRequest = SyncRequest();
+        syncRequest.onDone = _onDone;
+        syncRequest.onError = _onError;
+        syncRequest.requestSource = SyncRequestSource.system;
+        syncRequest.requestedOn = DateTime.now().toUtc();
+        syncRequest.result = SyncResult();
+        _syncRequests.add(syncRequest);
```

Comment:
Encapsulate in a method and these you decide whether to add or not and where to add